File: Daily/1143. Longest Common Subsequence/Solution1.py
from functools import lru_cache
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        '''
            Optimized Solution using Memoization
            Time Complexity: O(M*N)
            Space Complexity: O(M*N)
        '''
        '''
        text1_Size, text2_Size = len(text1) + 1, len(text2) + 1
        cache = [[-1] * text2_Size for _ in range(text1_Size)]
        
        @lru_cache(maxsize = None)
        def LCS(text1: str, text2: str) -> int:
            if len(text1) and len(text2):
                if cache[len(text1)][len(text2)] == -1:
                   if text1[0] == text2[0]: ans = 1 + LCS(text1[1:], text2[1:])
                   else: ans = max(LCS(text1, text2[1:]), LCS(text1[1:], text2)) 
                   cache[len(text1)][len(text2)] = ans
                return cache[len(text1)][len(text2)]
            return 0
        
        return LCS(text1, text2)
        '''
        @lru_cache(maxsize=None)
        def LCS(p1, p2):
            if p1 == len(text1) or p2 == len(text2):
                return 0
            if text1[p1] == text2[p2]:
                return 1 + LCS(p1 + 1, p2 + 1)
            else:
                return max(LCS(p1, p2 + 1), LCS(p1 + 1, p2))
        return LCS(0, 0)

if __name__ == '__main__':
    text1, text2 = tuple(input().split()); sol = Solution()
    print('Longest Common Subsequence:', sol.longestCommonSubsequence(text1, text2))

File: Daily/1143. Longest Common Subsequence/Solution1.cpp
#include <iostream>
#include <string>
#include <functional>
#include <vector>
using namespace std;
class Solution {
    public:
        int longestCommonSubsequence(string text1, string text2) {
            /*
                Optimized Solution using Recursion (Memoization)
                Time Complexity: O(M*N)
                Space Complexity: O(M*N)
            */
            vector<vector<int>> cache(text1.size()+1, vector<int>(text2.size()+1, -1));
            function<int(string, string)> LCS = [&](string text1, string text2) -> int {
                int text1_Size = text1.size(), text2_Size = text2.size();
                if(text1_Size && text2_Size) {
                    if(cache[text1_Size][text2_Size] == -1) {
                        int ans = 0;
                        if(text1[0] == text2[0]) ans = 1 + LCS(text1.substr(1), text2.substr(1));
                        else ans = max(LCS(text1.substr(1), text2), LCS(text1, text2.substr(1)));
                        cache[text1_Size][text2_Size] = ans;
                    }
                    return cache[text1_Size][text2_Size];
                }
                return 0;
            };
            return LCS(text1, text2);
        }
};
int main() {
    string text1, text2; Solution sol;
    cin >> text1 >> text2;
    cout << "Longest Common Subsequence: " << 
        sol.longestCommonSubsequence(text1, text2) << endl;
}